File: 数组/41.缺失的第一个正数.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        memo = set(nums)
        res = 1
        while True:
            if res in memo:
                res += 1
            else:
                return res

class Solution2:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        
        # 将每个数放到正确的位置上
        for i in range(n):
            while 1 <= nums[i] <= n and nums[nums[i] - 1] != nums[i]:
                nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
        
        # 找到第一个位置不正确的数
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1
        
        return n + 1

File: 数组/41.缺失的第一个正数.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function (nums) {
  let n = nums.length;

  // 将每个数放到正确的位置上
  for (let i = 0; i < n; i++) {
    let num = nums[i];
    //* 需要在正确的位置 && 数字num 对应索引num - 1
    while (1 <= num && num <= n && nums[num - 1] !== num) {
      //! num[i] 要后修改, nums[num - 1]使用到了 num[i]
      [nums[num - 1], num] = [num, nums[num - 1]];
    }
  }

  // 找到第一个位置不正确的数
  for (let i = 0; i < n; i++) {
    if (nums[i] !== i + 1) {
      return i + 1;
    }
  }

  return n + 1;
};
